service/src/account_funding.rs

```diff
@@ -16,50 +16,74 @@
 */
 
 use crate::error::{Error, ServiceResult};
-use itp_node_api::api_client::{AccountApi, ParentchainApi};
+use codec::Encode;
+use itp_api_client_types::ParentchainUncheckedExtrinsic;
+use itp_node_api::api_client::{AccountApi, ChainApi, UncheckedExtrinsicV4};
 use itp_settings::worker::{
 	EXISTENTIAL_DEPOSIT_FACTOR_FOR_INIT_FUNDS, REGISTERING_FEE_FACTOR_FOR_INIT_FUNDS,
 };
-use itp_types::parentchain::Balance;
+use itp_types::parentchain::{AccountId, Address, Balance, Hash, Index, Signature};
 use log::*;
 use sp_core::{
 	crypto::{AccountId32, Ss58Codec},
-	Pair,
+	sr25519, Pair,
 };
 use sp_keyring::AccountKeyring;
-use sp_runtime::MultiAddress;
+use sp_runtime::{MultiAddress, OpaqueExtrinsic};
 use substrate_api_client::{
-	extrinsic::BalancesExtrinsics, GetBalance, GetTransactionPayment, SubmitAndWatch, XtStatus,
+	extrinsic::{balances::TransferAllowDeathCall, BalancesExtrinsics},
+	GetBalance, GetTransactionPayment, SubmitAndWatch, XtStatus,
 };
 
 /// Information about the enclave on-chain account.
 pub trait EnclaveAccountInfo {
 	fn free_balance(&self) -> ServiceResult<Balance>;
 }
 
-pub struct EnclaveAccountInfoProvider {
+pub struct EnclaveAccountInfoProvider<ParentchainApi> {
 	node_api: ParentchainApi,
 	account_id: AccountId32,
 }
 
-impl EnclaveAccountInfo for EnclaveAccountInfoProvider {
+impl<ParentchainApi> EnclaveAccountInfo for EnclaveAccountInfoProvider<ParentchainApi>
+where
+	ParentchainApi: AccountApi<AccountId = AccountId, Balance = Balance>,
+{
 	fn free_balance(&self) -> ServiceResult<Balance> {
 		self.node_api.get_free_balance(&self.account_id).map_err(|e| e.into())
 	}
 }
 
-impl EnclaveAccountInfoProvider {
+impl<ParentchainApi> EnclaveAccountInfoProvider<ParentchainApi> {
 	pub fn new(node_api: ParentchainApi, account_id: AccountId32) -> Self {
-		EnclaveAccountInfoProvider { node_api, account_id }
+		Self { node_api, account_id }
 	}
 }
 
-pub fn setup_account_funding(
+pub fn setup_account_funding<ParentchainApi>(
 	api: &ParentchainApi,
 	accountid: &AccountId32,
 	encoded_extrinsic: Vec<u8>,
 	is_development_mode: bool,
-) -> ServiceResult<()> {
+) -> ServiceResult<()>
+where
+	ParentchainApi: AccountApi<AccountId = AccountId, Balance = Balance, Index = Index>
+		+ GetBalance<Balance = Balance>
+		+ GetTransactionPayment<Balance = Balance>
+		+ BalancesExtrinsics<
+			Address = Address,
+			Balance = Balance,
+			Extrinsic<TransferAllowDeathCall<Address, Balance>> = UncheckedExtrinsicV4<
+				Address,
+				TransferAllowDeathCall<Address, Balance>,
+				<ParentchainApi as ChainApi>::Signature,
+				<ParentchainApi as ChainApi>::SignedExtra,
+			>,
+		> + SubmitAndWatch<Hash = Hash>
+		+ ChainApi<Signer = sr25519::Pair>,
+	ParentchainApi::Signature: Encode,
+	ParentchainApi::SignedExtra: Encode,
+{
```

Comment:
this is so horribly verbose. there has to be another way

Reply:
You can create an aggregate trait like here: https://stackoverflow.com/a/26983395/19331449